### ISSUE TITLE
SegFault when connection is failing on mobile network (MQTT / WS-MQTT)

### ISSUE THREAD

**Issue by pboettch** [Azure/azure-iot-sdk-c#459]
- **OS and version used:** Debian Stretch x86-x64
- **SDK version used:**  2018-03-02

# Description of the issue:

My Azure connection is established via mobile network (2G, 3G or 4G) which comes and goes because my physical device host is moving (it's a train).

From time to time my application is crashing, when a connection could not be established. When testing it manually I realized that before a crash I always see the same debug-print in the console (more below).

# Code sample exhibiting the issue:
My code is C++ using the C-SDK. I'm using the low-level-API and call do-work-from a `std::thread`:

```C++
while (!stop_ || fadingOutCount) {
	if (stop_) {
		fadingOutCount--;
	}

	IOTHUB_CLIENT_STATUS status;
	auto ret = IoTHubClient_LL_GetSendStatus(iotHubClientHandle_, &status);
	if (ret != IOTHUB_CLIENT_OK)
		error("GetSendStatus is not OK {}", ret);
	else if (status != prev) {
		debug("SendStatus before dowork changed: - busy {}", status);
		prev = status;
	}

	IoTHubClient_LL_DoWork(iotHubClientHandle_);

	std::this_thread::sleep_for(50msec);
}
```

This is my init-code:

```C++
bool init()
{
	if (platform_init() != 0) {
		log("Azure Platform init failed.");
		return false;
	}

	setConnectionStatus(IOTHUB_CLIENT_CONNECTION_UNAUTHENTICATED, IOTHUB_CLIENT_CONNECTION_OK);

	iotHubClientHandle_ = IoTHubClient_LL_CreateFromConnectionString(connectionString_.c_str(),
																	 MQTT_Protocol);
	if (!iotHubClientHandle_) {
		log("L_MC_R_999", {"Could not create handle from connection string", connectionString_});
		return false;
	}

	bool traceOn = false;
	auto ret = IoTHubClient_LL_SetOption(iotHubClientHandle_, OPTION_LOG_TRACE, &traceOn);
	if (ret != IOTHUB_CLIENT_OK)
		log("L_MC_R_999", {"SetOption logtrace failed"});

	size_t lifetime = 24 * 3600;
	ret = IoTHubClient_LL_SetOption(iotHubClientHandle_, OPTION_SAS_TOKEN_LIFETIME, &lifetime);
	if (ret != IOTHUB_CLIENT_OK)
		log("L_MC_R_999", {"SetOption SAS_TOKEN_LIFETIME failed"});

	int keepalive = 6 * 60;
	ret = IoTHubClient_LL_SetOption(iotHubClientHandle_, OPTION_KEEP_ALIVE, &keepalive);
	if (ret != IOTHUB_CLIENT_OK)
		log("L_MC_R_999", {"SetOption KEEP_ALIVE failed"});

	// does not work in MQTT
	//ret = IoTHubClient_LL_SetOption(iotHubClientHandle_, OPTION_SAS_TOKEN_REFRESH_TIME, &lifetime);
	//if (ret != IOTHUB_CLIENT_OK)
	//	log("L_MC_R_999", {"SetOption SAS_TOKEN_REFRESH_TIME failed"});

	// has no effect
	//tickcounter_ms_t timeout_msg = MESSAGE_TIMEOUT;
	//ret = IoTHubClient_LL_SetOption(iotHubClientHandle_, OPTION_MESSAGE_TIMEOUT", &timeout_msg);
	//if (ret != IOTHUB_CLIENT_OK)
	//	log("L_MC_R_999", {"SetOption messageTimeout failed"});

	// not used
	//ret = IoTHubClient_LL_SetMessageCallback(
	//    iotHubClientHandle_,
	//    [](IOTHUB_MESSAGE_HANDLE message,
	//       void *ctx) { return reinterpret_cast<AzureConnection *>(ctx)->handleMessage(message); },
	//    this);
	//if (ret != IOTHUB_CLIENT_OK) {
	//	log("L_MC_R_999", {"IoTHubClient_LL_SetMessageCallback failed!"});
	//	return false;
	//}

	ret = IoTHubClient_LL_SetDeviceMethodCallback_Ex(
		iotHubClientHandle_,
		[](const char *method_name,
		   const unsigned char *payload,
		   size_t size,
		   METHOD_HANDLE method_id,
		   void *ctx) {
			return reinterpret_cast<AzureConnection *>(ctx)->inboundDeviceMethod(method_name, payload, size, method_id);
		},
		this);
	if (ret != IOTHUB_CLIENT_OK) {
		log("L_MC_R_999", {"IoTHubClient_LL_SetDeviceMethodCallback_Ex failed!"});
		return false;
	}

	ret = IoTHubClient_LL_SetConnectionStatusCallback(
		iotHubClientHandle_,
		[](IOTHUB_CLIENT_CONNECTION_STATUS result,
		   IOTHUB_CLIENT_CONNECTION_STATUS_REASON reason,
		   void *ctx) { reinterpret_cast<AzureConnection *>(ctx)->setConnectionStatus(result, reason); },
		this);
	if (ret != IOTHUB_CLIENT_OK) {
		log("L_MC_R_999", {"IoTHubClient_LL_SetConnectionStatusCallback failed!"});
		return false;
	}

	return true;
}
```

(Sorry for the lengthy code dump)

# Console log of the issue:

As mentioned above the seg-fault, if it appears, always is followed by the following debug-line:

```
Error: Time:Mon Apr 16 07:39:31 2018 File:[..]/iothub_client/src/iothubtransport_mqtt_common.c Func:InitializeConnection Line:1855 mqtt_client timed out waiting for CONNACK
```

Preceeded by
```
Error: Time:Mon Apr 16 07:38:31 2018 File:[..]/c-utility/adapters/socketio_berkeley.c Func:socketio_open Line:682 Failure: select failure.
Error: Time:Mon Apr 16 07:38:31 2018 File:[..]/c-utility/src/tlsio_openssl.c Func:on_underlying_io_open_complete Line:751 Invalid tlsio_state. Expected state is TLSIO_STATE_OPENING_UNDERLYING_IO.
Error: Time:Mon Apr 16 07:38:31 2018 File:[..]/c-utility/src/tlsio_openssl.c Func:tlsio_openssl_open Line:1251 Failed opening the underlying I/O.
Error: Time:Mon Apr 16 07:38:31 2018 File:[..]/umqtt/src/mqtt_client.c Func:mqtt_client_connect Line:980 Error: io_open failed
Error: Time:Mon Apr 16 07:38:31 2018 File:[..]/iothub_client/src/iothubtransport_mqtt_common.c Func:SendMqttConnectMsg Line:1787 failure connecting to address XXX.azure-devices.net.
```

These lines are appearing always when a connection can't be established and there is no segfault. The segfault only happens when there is the "timed out waiting for CONNACK"-line.

I'm able to capture the crash within GDB (a session which I still have available for further debugging if needed), here's the backtrace:

```
#0  0x00005555556301e4 in xio_dowork (xio=0x7fffe8053ae0) at .../external/azure/azure-iot-sdk-c/c-utility/src/xio.c:165
#1  0x000055555562a398 in mqtt_client_dowork (handle=0x5555558d1170) at .../external/azure/azure-iot-sdk-c/umqtt/src/mqtt_client.c:1227
#2  0x00005555556255fe in IoTHubTransport_MQTT_Common_DoWork (handle=0x5555558cb0d0, iotHubClientHandle=0x5555559063b0) at .../external/azure/azure-iot-sdk-c/iothub_client/src/iothubtransport_mqtt_common.c:2612
#3  0x000055555561f622 in IoTHubTransportMqtt_DoWork (handle=0x5555558cb0d0, iotHubClientHandle=0x5555559063b0) at .../external/azure/azure-iot-sdk-c/iothub_client/src/iothubtransportmqtt.c:115
#4  0x0000555555612b86 in IoTHubClient_LL_DoWork (iotHubClientHandle=0x5555559063b0) at .../external/azure/azure-iot-sdk-c/iothub_client/src/iothub_client_ll.c:1325
#5  0x00005555555cfaeb in (anonymous namespace)::AzureConnection::process (this=0x7fffffffd7f0) at ../app/router-azure-client.cpp:358
#6  0x00005555555d46dc in std::__invoke_impl<void, void ((anonymous namespace)::AzureConnection::*)(), (anonymous namespace)::AzureConnection*>(std::__invoke_memfun_deref, <unknown type in .../build/app/router-azure-client, CU 0x0, DIE 0x84c4c>, <unknown type in .../build/app/router-azure-client, CU 0x0, DIE 0x84c5a>) ( __f=<unknown type in .../build/app/router-azure-client, CU 0x0, DIE 0x84c4c>, __t=<unknown type in .../build/app/router-azure-client, CU 0x0, DIE 0x84c5a>) at /usr/include/c++/7/bits/invoke.h:73
#7  0x00005555555d3a8a in std::__invoke<void ((anonymous namespace)::AzureConnection::*)(), (anonymous namespace)::AzureConnection*>(<unknown type in .../build/app/router-azure-client, CU 0x0, DIE 0x 8c497>, <unknown type in .../build/app/router-azure-client, CU 0x0, DIE 0x8c4b2>) (__fn=<unknown type in .../build/app/router-azure-client, CU 0x0, DIE 0x8c497>, __args#0=<unknown type in .../build/app/router-azure-client, CU 0x0, DIE 0x8c4b2>) at /usr/include/c++/7/bits/invoke.h:95
#8  0x00005555555d63b5 in std::thread::_Invoker<std::tuple<void ((anonymous namespace)::AzureConnection::*)(), (anonymous namespace)::AzureConnection*> >::_M_invoke<0, 1> (this=0x555555906858) at /usr/include/c++/7/thread:234
#9  0x00005555555d6347 in std::thread::_Invoker<std::tuple<void ((anonymous namespace)::AzureConnection::*)(), (anonymous namespace)::AzureConnection*> >::operator() (this=0x555555906858) at /usr/include/c++/7/thread:243
#10 0x00005555555d6306 in std::thread::_State_impl<std::thread::_Invoker<std::tuple<void ((anonymous namespace)::AzureConnection::*)(), (anonymous namespace)::AzureConnection*> > >::_M_run (this=0x555555906850) at /usr/include/c++/7/thread:186
#11 0x00007ffff6342843 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#12 0x00007ffff66125aa in start_thread (arg=0x7fffefb4d700) at pthread_create.c:463
#13 0x00007ffff5a17cbf in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

The line of code is as follows:

```
xio_instance->io_interface_description->concrete_io_dowork(xio_instance->concrete_xio_handle);
```

I checked all of the references, nothing is null.

I could reproduce the problem also with the February release and using WebSocket+MQTT.

Any help is appreciated.

**Comment by villepalo** [comment]
Seems to be the same issue as #446

**Comment by pboettch** [comment]
I'm leaving this one open until I tried the fix.

**Comment by pboettch** [comment]
Fix is discussed in #466.

**Comment by tameraw** [comment]
@pboettch @villepalo thank you for your contribution to our open-sourced project!  Please help us improve by filling out this 2-minute [customer satisfaction survey](https://www.surveymonkey.com/r/Q5PLZ3V).
